### DISCUSSION TITLE
Compiling on Chimera

### DISCUSSION THREAD

**Discussion by NovaAndrom3da** [hyprwm/Hyprland#3101] [Q&A] [answered]
Finally, after the whole fiasco with gitlab.fd.o being down, I'm actually able to get somewhere with compiling. I'm currently trying to compile on [Chimera Linux](https://chimera-linux.org) which is a musl/clang/llvm system. I have libegl, libgles1, libgles2, mesa, glu, gegl, freeglut, libgl, glew and all of their development packages installed (with the exception of libegl, which doesn't have a devel package- maybe part of my problem?). I get through a compile mostly successfully. I'd also like to mention that these lines are also a part of the output:

```
Dependency egl found: YES 23.1.6 (cached)
[...]
Dependency libdrm found: YES 2.4.116 (cached)
```

```
wlroots 0.17.0-dev

    drm-backend     : NO
    [blah blah blah]
    [xcb-errors is missing but that's not important]
    egl             : YES
```
(I included the drm message just cause that's odd but not what I suspect to be a root cause of my issue)
Now with that being said, even though the egl libraries are available, it complains about not being able to find it:

```
-- Setting link libraries
-- Configuring done (1.3s)
CMake Error at CMakeLists.txt:195 (target_link_libraries)
  Target "Hyprland" links to:

    OpenGL::EGL

  but the target was not found. Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.



-- Generating done (0.0s)
```

It seems to continue (with a bunch of errors (`'execinfo.h' file not found`, and a few others complaining about C-linkage being specified but returning incomplete type 'std::[stuff]') if I comment line 197 of ./CMakeLists.txt (`OpenGL::EGL` in `target_link_libraries`). Any ideas?

EDIT: Forgot to mention- I can upload the whole file (if I can figure out hoe to, currently without a GUI on my laptop, so things are a bit rough lol)

**Comment by memchr** [comment, marked as answer]
> musl/clang/llvm system

Hyprland uses several C++23 features that Clang either does not implement or has bugs in, such as ranges. So you will need gcc to compile this project, if you do not already have it.

>  I can upload the whole file (if I can figure out hoe to, currently without a GUI on my laptop, so things are a bit rough lol)

There are paste services that can be used with curl such as ix.io or paste.rs.

```bash
<command> | curl -F 'f:1=<-' ix.io
# or
curl -F 'f:1=<-' ix.io < /path/to/file
```

---

Also, do you have any luck with meson?

```bash
meson setup --buildtype=release build
# Limited to 6 concurrent jobs as Ninja has unsafe settings at the moment
meson compile -j6 -C build
```

**Reply by NovaAndrom3da** [reply]
Yes, meson seems to be a bit better, and a lot of the errors have been downgraded to warnings. Seems like my only issue now is missing `execinfo.h`

**Reply by NovaAndrom3da** [reply]
http://ix.io/4ENo

**Reply by NovaAndrom3da** [reply]
I think execinfo might be something I'll have to figure out on the chimera side, thank you 🤗

**Reply by memchr** [reply]
`execinfo.h` is a glibc-only header, you may need to modify the hyprland source, and possibly not just this execinfo issue, to get it to compile with musl.

**Reply by memchr** [reply]
Since backtrace and backtrace_symbols from execinfo.h are only used for crashreporting, you may remove the inclusion of this header. 

Can you try this patch
```diff
diff --git a/src/debug/CrashReporter.cpp b/src/debug/CrashReporter.cpp
index a580ecd..c0c683d 100644
--- a/src/debug/CrashReporter.cpp
+++ b/src/debug/CrashReporter.cpp
@@ -1,7 +1,9 @@
 #include "CrashReporter.hpp"
 #include <random>
 #include <sys/utsname.h>
+#ifdef __GLIBC__
 #include <execinfo.h>
+#endif
 #include <fstream>
 #include <signal.h>
 
diff --git a/src/helpers/MiscFunctions.cpp b/src/helpers/MiscFunctions.cpp
index 7bf254e..1256e97 100644
--- a/src/helpers/MiscFunctions.cpp
+++ b/src/helpers/MiscFunctions.cpp
@@ -6,7 +6,9 @@
 #include <sys/utsname.h>
 #include <iomanip>
 #include <sstream>
+#ifdef __GLIBC__
 #include <execinfo.h>
+#endif
 
 #if defined(__DragonFly__) || defined(__FreeBSD__) || defined(__NetBSD__) || defined(__OpenBSD__)
 #include <sys/sysctl.h>
@@ -699,6 +701,7 @@ std::string replaceInString(std::string subject, const std::string& search, cons
 }
 
 std::vector<SCallstackFrameInfo> getBacktrace() {
+#ifdef __GLIBC__
     std::vector<SCallstackFrameInfo> callstack;
 
     void*                            bt[1024];
@@ -711,8 +714,10 @@ std::vector<SCallstackFrameInfo> getBacktrace() {
     for (size_t i = 0; i < btSize; ++i) {
         callstack.emplace_back(SCallstackFrameInfo{bt[i], std::string{btSymbols[i]}});
     }
-
     return callstack;
+#else
+    return std::vector<SCallstackFrameInfo>();
+#endif
 }
 
 void throwError(const std::string& err) {
```

**Reply by NovaAndrom3da** [reply]
Yeah, I already patched all that out :)
